### PR TITLE
executor: forbid user to drop important system table

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

If a user drop some important tables by mistake, such as mysql.user, it's difficult to recover.
In the worst case, the whole cluster fails to bootstrap, we should prevent that.

### What is changed and how it works?

Forbid user to drop important system tables.

### Check List 

Tests 

 - Unit test

Side effects

It's harder to drop those tables.

Should we also protect the statistic tables?  @lamxTyler 
@zimulala @shenli

### CODE REVIEW

File: executor/ddl.go [hunk 1]
```diff
@@ -154,8 +155,19 @@ func (e *DDLExec) executeCreateIndex(s *ast.CreateIndexStmt) error {
...
 
+var errForbidDrop = errors.New("Hey, drop system database/table, are you sure?")
+
 func (e *DDLExec) executeDropDatabase(s *ast.DropDatabaseStmt) error {
 	dbName := model.NewCIStr(s.Name)
+
+	// Protect important system table from been dropped by a mistake.
+	// I can hardly find a case that a user really need to do this.
+	if dbName.L == "mysql" {
+		if !privileges.SkipWithGrant {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/ddl.go]: I know it's a rare case, but how about that we really want to do it?

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/ddl.go]: We need a "backdoor".

File: executor/ddl.go [hunk 2]
```diff
@@ -198,6 +230,14 @@ func (e *DDLExec) executeDropTable(s *ast.DropTableStmt) error {
 			return errors.Trace(err)
 		}
 
+		// Protect important system table from been dropped by a mistake.
+		// I can hardly find a case that a user really need to do this.
+		if isSystemTable(tn.Schema.L, tn.Name.L) {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/ddl.go]: What's your opinion? @shenli @lamxTyler

[the remark this answers is not in the document]
Comment (shenli) [reply executor/ddl.go]: I suggest protecting all the tables in the `mysql` database.

File: executor/ddl.go [hunk 3]
```diff
@@ -154,8 +155,19 @@ func (e *DDLExec) executeCreateIndex(s *ast.CreateIndexStmt) error {
 	return errors.Trace(err)
 }
 
+var errForbidDrop = errors.New("Hey, drop system database/table, are you sure?")
```
Comment (shenli) [inline executor/ddl.go]: Please make the error message more seriously.

File: executor/ddl.go [hunk 4]
```diff
@@ -179,6 +191,26 @@ func (e *DDLExec) executeDropDatabase(s *ast.DropDatabaseStmt) error {
...
+var systemTables = map[string]struct{}{
+	"user":                 {},
+	"db":                   {},
+	"tables_priv":          {},
+	"columns_priv":         {},
+	"gc_delete_range":      {},
+	"gc_delete_range_done": {},
+}
+
+func isSystemTable(schema, table string) bool {
```
Comment (zimulala) [inline executor/ddl.go]: Could we put it to infoschema.go? I think it is like "IsMemoryDB", we can put these together.

Comment (tiancaiamao) [reply executor/ddl.go]: This function is removed, we forbid dropping tables in "mysql" database.

**Comment by tiancaiamao** [comment]
A sane person will not drop system tables.
A sane person will not create table on "mysql" database
This change tries to prevent some insane person from breaking the system, it has very small affect on a sane person.

**Comment by jackysp** [comment]
"A sane person will not create table on "mysql" database" ? Maybe s/sane/genius/ will be all right :-)

**Comment by shenli** [comment]
@morgo Some tables in the `mysql` database are critical to TiDB. For example `tidb`, 'gc_*'. Users should never touch these tables.

@tiancaiamao `User`, '*_Priv' tables are not so critical. We could save the cluster even they are dropped. I think we could relax the protection for them.

**Comment by tiancaiamao** [comment]
PTAL @shenli @jackysp

**Comment by tiancaiamao** [comment]
PTAL @shenli

**Comment by tiancaiamao** [comment]
PTAL @shenli

File: executor/ddl.go [hunk 5]
```diff
@@ -198,6 +223,12 @@ func (e *DDLExec) executeDropTable(s *ast.DropTableStmt) error {
 			return errors.Trace(err)
 		}
 
+		// Protect important system table from been dropped by a mistake.
+		// I can hardly find a case that a user really need to do this.
+		if isSystemTable(tn.Schema.L, tn.Name.L) {
+			return errors.Trace(errForbidDrop)
```
Comment (shenli) [inline executor/ddl.go]: Add the system table name in the error message.

File: executor/ddl.go [hunk 6]
```diff
@@ -179,6 +188,22 @@ func (e *DDLExec) executeDropDatabase(s *ast.DropDatabaseStmt) error {
 	return errors.Trace(err)
 }
 
+var systemTables = map[string]struct{}{
```
Comment (shenli) [inline executor/ddl.go]: Add a comment for this map. Explain why we prevent users from dropping them.

**Comment by tiancaiamao** [comment]
PTAL @shenli 
/run-all-tests

**Pull request merged**